src/backend/opencl/compile_module.cpp

```diff
@@ -141,10 +202,45 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
 
 Module loadModuleFromDisk(const int device, const string &moduleKey,
                           const bool isJIT) {
-    UNUSED(device);
-    UNUSED(moduleKey);
-    UNUSED(isJIT);
-    return {};
+    const string &cacheDirectory = getCacheDirectory();
+    if (cacheDirectory.empty()) return Module{};
+
+    auto &dev              = opencl::getDevice(device);
+    const string cacheFile = cacheDirectory + AF_PATH_SEPARATOR +
+                             getKernelCacheFilename(device, moduleKey);
+    try {
+        std::ifstream in(cacheFile, std::ios::binary);
+        if (!in.is_open()) {
+            AF_ERROR("Unable to open binary cache file", AF_ERR_INTERNAL);
+        }
+        in.exceptions(std::ios::failbit | std::ios::badbit);
+
+        // TODO Handle cases where program objects are created from contexts
+        // having multiple devices
+        size_t clbinHash = 0;
+        in.read(reinterpret_cast<char *>(&clbinHash), sizeof(clbinHash));
+        size_t clbinSize = 0;
+        in.read(reinterpret_cast<char *>(&clbinSize), sizeof(clbinSize));
+        vector<unsigned char> clbin(clbinSize);
+        in.read(reinterpret_cast<char *>(clbin.data()), clbinSize);
+        in.close();
```

Comment:
nit: This will close when `in` goes out of scope

Reply:
true again, but there are quite a few operations going on after reading the file contents, recomputation of hash, cl::program construction and there is no reason to keep the stream open while doing those.